File: apps/ios/PassAlarm/Domain/UseCase/OnAlarmFiredUseCase.swift
import Foundation

final class OnAlarmFiredUseCase: Sendable {
    private let tokenRepository: ScheduledTokenRepositoryProtocol
    private let reschedule: RescheduleNextNUseCase

    init(tokenRepository: ScheduledTokenRepositoryProtocol,
         reschedule: RescheduleNextNUseCase) {
        self.tokenRepository = tokenRepository
        self.reschedule = reschedule
    }

    func execute(osIdentifier: String) async {
        do {
            if let token = try await tokenRepository.fetchByOsIdentifier(osIdentifier) {
                try await tokenRepository.updateStatus(token.id, status: .fired)
            }
            try await reschedule.execute()
        } catch {
            print("OnAlarmFiredUseCase error: \(error)")
        }
    }
}

File: apps/android/app/src/main/kotlin/com/nogeass/passalarm/domain/usecase/OnAlarmFiredUseCase.kt
package com.nogeass.passalarm.domain.usecase

import com.nogeass.passalarm.domain.model.TokenStatus
import com.nogeass.passalarm.domain.repository.ScheduledTokenRepository
import javax.inject.Inject

class OnAlarmFiredUseCase @Inject constructor(
    private val tokenRepository: ScheduledTokenRepository,
    private val reschedule: RescheduleNextNUseCase
) {
    suspend fun execute(osIdentifier: Int) {
        val token = tokenRepository.fetchByOsIdentifier(osIdentifier) ?: return
        tokenRepository.updateStatus(token.id, TokenStatus.FIRED)
        reschedule()
    }
}
